### PR TITLE
Added the beginning of a block "material" component.

### PR DESCRIPTION
**Pull request by winsock**
This was needed in-order to enable the playing of step-sounds and is needed to close NOVA-Team/NOVA-Wrapper-MC#21

### CODE REVIEW

File: src/main/java/nova/core/component/misc/BlockMaterial.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	public boolean requiresNoTool() {
+		return requiresNoTool;
+	}
+
+	public BlockMaterial setRequiresNoTool(boolean requiresNoTool) {
+		this.requiresNoTool = requiresNoTool;
+		return this;
+	}
+
+	public void setBlocksMovement(boolean blocksMovement) {
```
Comment (anti344) [inline src/main/java/nova/core/component/misc/BlockMaterial.java]: Make this chainable too?
And all below

File: src/main/java/nova/core/block/Block.java [hunk 2]
```diff
@@ -52,6 +53,10 @@ public final BlockTransform transform() {
 		return get(BlockTransform.class);
 	}
 
+	public final BlockMaterial material() {
+		return getOrAdd(new BlockMaterial());
```
Comment (RX14) [inline src/main/java/nova/core/block/Block.java]: Other components don't have this...

**Comment by RX14** [comment]
I feel this needs more discussion as to whether a more abstract system can be accomplished, this seems really MC specific...

File: src/main/java/nova/core/block/Block.java (hunk 2 above)
Comment (winsock) [inline src/main/java/nova/core/block/Block.java]: Blocks need a material in the super constructor(in MC at least), so it's either this or an null check at the other end.

Comment (RX14) [inline src/main/java/nova/core/block/Block.java]: you should use `block.getOrAdd(new BlockMaterial())` on the other end.

File: src/main/java/nova/core/component/misc/BlockMaterial.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+		public BlockSound(String type) {
+			super();
+			this.type = type;
+		}
+
+		@Override
+		public String getID() {
+			if (blockSoundResource.contains(":")) {
+				return blockSoundResource;
```
Comment (RX14) [inline src/main/java/nova/core/component/misc/BlockMaterial.java]: Aren't these sounds MC specific?

**Comment by winsock** [comment]
I feel like it may be possible to get a bit more abstraction. The naming as it stands is based off of the MC code, so that will look MC specific till better names are found, however, wouldn't most voxel games have these properties in one form or another?

**Comment by calclavia** [comment]
I'm wondering if BlockMaterial is a component that's universal to any game. This looks VERY MC-specific. We could separate BlockMaterial into different components...

File: src/main/java/nova/core/component/misc/BlockMaterial.java (hunk 3 above)
Comment (winsock) [inline src/main/java/nova/core/component/misc/BlockMaterial.java]: This is a gray area IMO. The idea of sounds associated with blocks would be agnostic AFAIK. There might be a better solution here, however I didn't see it.

**Comment by RX14** [comment]
This is basically metadata, most of the time this component would be a no-op in games, is that acceptable?

**Comment by RX14** [comment]
@calclavia I do agree with seperating into other components though

**Comment by calclavia** [comment]
@RX14 If it's MC specific, then it should go into the MC plugin. But I DO think this will be applicable in many games. Block break sound, for example, seems like an intrinsic property that Block should have, not based on its material...

**Comment by winsock** [comment]
@calclavia @RX14 I agree as well. I was wanting to avoid editing the Block class, however I think most of these would fit into separate components and/or some in the block class itself.

**Comment by RX14** [comment]
We really want the block class to be bare bones, components are better.

File: src/main/java/nova/core/component/misc/BlockMaterial.java (hunk 3 above)
Comment (RX14) [inline src/main/java/nova/core/component/misc/BlockMaterial.java]: The issue is that the defaults, "stone" may not exist and the naming of "step.stone" and "dig.stone" may not exist either.

**Comment by winsock** [comment]
I think that all of the properties in BlockMaterial are intrinsic to most Voxel engine games that use a "block" syle world. I would think they would feel incomplete if they didn't have their own use of these properties. In the end it's just a container class for more "advanced" block properties, i.e. user interactive/immersive blocks.

**Comment by anti344** [comment]
I feel like other voxel games might not have block resistance, only hardness.
Component, maybe? =D

**Comment by calclavia** [comment]
@anti344 I agree. Explodable Component would be better.

**Comment by calclavia** [comment]
Maybe the light properties should go into LightEmiter component? Refactor the existing LightEmtiter to LightHandler.

**Pull request merged**